lib/kontena/actors/haproxy_config_generator.rb

```diff
@@ -0,0 +1,121 @@
+module Kontena::Actors
+  class HaproxyConfigGenerator < Concurrent::Actor::RestartingContext
+    include Kontena::Logging
+
+    def initialize
+      info "initialized"
+    end
+
+    # @param [Message] msg
+    def on_message(msg)
+      case msg.action
+      when :update
+        update(msg.value)
+      else
+        pass
+      end
+    end
+
+    # @param [Etcd::Node] node
+    def update(node)
+      root = node.key
+      services = []
+      tcp_services = []
+      node.children.each do |c|
+        if c.key == "#{root}/services"
+          services = generate_services(c)
+        elsif c.key == "#{root}/tcp-services"
+          tcp_services = generate_tcp_services(c)
+        end
+      end
+
+      config = Kontena::Views::Haproxy.render({
+        format: :text, services: services, tcp_services: tcp_services
+      }).each_line.reject{ |l| l.strip == ''.freeze }.join
+      parent << Message.new(:write_config, config)
```

Comment:
Hmm... if the models are just rendered out to a string, do they need to get frozen?

Reply:
Not sure what you mean?

Reply:
I initially assumed that the point of the model `freeze`s in the `generate_service` etc functions was to prepare the model objects for inclusion in messages sent between actors, but if the models get rendered to a string in the same actor, then that's moot.

I guess they're just to keep the view templates from mutating the models?

Reply:
Yeah, I'm not 💯 sure that rendering them in the same actor makes actually sense. Probably rendering should be done somewhere else (would make pluggable datasource easy). IMHO this refactoring can be done later.